pychromecast/controllers/spotify.py

```diff
@@ -0,0 +1,169 @@
+
+        response = session.get("https://accounts.spotify.com/login",
+                               headers=headers, cookies=cookies)
+        response.raise_for_status()
+
+        return response.cookies['csrf_token']
+
+    # pylint: disable=too-many-arguments
+    def _login(self, session, cookies, username, password, csrf_token):
+        """ Logs in with CSRF token and cookie within session. """
+        headers = {'user-agent': USER_AGENT}
+
+        data = {"remember": False, "username": username, "password": password,
+                "csrf_token": csrf_token}
+
+        response = session.post("https://accounts.spotify.com/api/login",
+                                data=data, cookies=cookies, headers=headers)
+
+        response.raise_for_status()
+
+    def _get_access_token(self, session, cookies):
+        """ Gets access token after login has been successful. """
+        headers = {'user-agent': USER_AGENT}
+
+        response = session.get("https://open.spotify.com/browse",
+                               headers=headers, cookies=cookies)
+        response.raise_for_status()
+
+        self.token = response.cookies['wp_access_token']
+
+        expiration = response.cookies['wp_expiration']
+        self.expiration_date = int(expiration)//1000
+
+    # Access token provided by spotipy does not have enough permissions to start
+    # Spotify on Chromecast
+    def start_session(self):
+        """ Starts session to get access token. """
+
+        # arbitrary value and can be static
+        cookies = {"__bon": "MHwwfC01ODc4MjExMzJ8LTI0Njg4NDg3NTQ0fDF8MXwxfDE="}
+
+        if self.username is None:
+            self.username = os.getenv("SPOTIFY_USERNAME")
+
+        if self.password is None:
+            self.password = os.getenv("SPOTIFY_PASS")
+
+        if self.username is None or self.password is None:
+            raise Exception("No username or password")
+
+        session = requests.Session()
+        token = self. _get_csrf(session, cookies)
+
+        self._login(session, cookies, self.username, self.password, token)
+        self._get_access_token(session, cookies)
+
+        self.session_started = True
+
+        return self.token
+
+    def launch_app(self):
+        """ Launch main application """
+
+        def callback():
+            """Callback function"""
+            self.send_message({"type": TYPE_STATUS, "credentials": self.token})
+
+        curr_time = time.time()
+
+        if self.session_started and curr_time < self.expiration_date:
+            self.logger.debug("Using same token: %s", self.token)
+            self.launch(callback_function=callback)
+        else:
+            self.logger.debug("Creating new token")
+            self.start_session()
+            self.logger.debug("Token is: %s", self.token)
+            self.launch(callback_function=callback)
+
+        # Need to wait for Spotify to be launched on Chromecast completely
+        time.sleep(5)
```

Comment:
Can't we poll here, maybe even using `receive_message` or your callback defined above? Using `time.sleep` seems like quite the hack to me...

Reply:
That might be much better. Any pointers on how to poll this? Not sure how to poll the socket_client for when the app is fully initialized?